geiger/src/graph/NodeWithPort.js

```diff
@@ -0,0 +1,282 @@
+    textSandbox.setText(this.nodeLabel);
+    maxTextLength = Math.max(
+        maxTextLength, textSandbox.clientWidth());
+    textSandbox.stop('nodeText');
+
+    return maxTextLength;
+  }
+
+  /**
+   * Setup the properties of each input port based on the message.
+   * @param {!NodeCreationMessage} message
+   * @param {?NodeDimension} dimension
+   * @return {Array<Port>}
+   */
+  _prepareInputPorts(message, dimension) {
+    const ports = [];
+    let inputPortY = INPUT_PORT_RADIUS + LEFT_SIDE_TOP_PADDING;
+    for (let i = 0; i < message.numberOfInputs; i++) {
+      ports.push({
+        'id': computeInPortId(message.contextId, message.nodeId, i),
+        'type': PortTypes.IN,
+        'text': '' + i,
+        'y': inputPortY,
+      });
+      inputPortY += TOTAL_INPUT_PORT_HEIGHT;
+    }
+    return ports;
+  }
+
+  /**
+   * Setup the properties of each output port based on the message.
+   * @param {!NodeCreationMessage} message
+   * @param {?NodeDimension} dimension
+   * @return {Array<Port>}
+   */
+  _prepareOutputPorts(message, dimension) {
+    const ports = [];
+    for (let i = 0; i < message.numberOfOutputs; i++) {
+      ports.push({
+        'id': computeOutPortId(message.contextId, message.nodeId, i),
+        'type': PortTypes.OUT,
+        'text': '' + i,
+      });
+    }
+    return ports;
+  }
+
+  /**
+   * Setup the properties of each param port based on the message.
+   * @param {!NodeCreationMessage} message
+   * @param {!NodeDimension} dimension
+   * @return {Array<Port>}
+   */
+  _prepareParamPorts(message, dimension) {
+    const ports = [];
+    let audioParamY = dimension.inputPortSectionHeight +
+        LEFT_SIDE_TOP_PADDING + AUDIO_PARAM_RADIUS;
+    for (let i = 0; i < message.audioParamNames.length; i++) {
+      ports.push({
+        'id': computeAudioParamPortId(
+            message.contextId, message.nodeId, message.audioParamNames[i]),
+        'type': PortTypes.PARAM,
+        'text': message.audioParamNames[i],
+        'y': audioParamY,
+      });
+      audioParamY += TOTAL_PARAM_PORT_HEIGHT;
+    }
+
+    return ports;
+  }
+
+  /**
+   * Validate ports created and then store.
+   * @param {!Array<Port>} ports
+   * @return {!Collection}
+   */
+  _createPortData(ports) {
+    const err = this._validatePorts(ports);
+    if (err.length) {
+      throw new Error(err);
+    }
+
+    return new Collection(ports);
+  }
+
+  /**
+   * Ensure each port is an object and has an id.
+   * @param {Array<Port>} ports
+   * @return {Array<string>} errors
+   */
+  _validatePorts(ports) {
+    const errorMessages = [];
+    ports.forEach((p) => {
+      if (typeof p !== 'object') {
+        errorMessages.push('Invalid port', p);
+      }
+
+      if (!this._isValidPortId(p.id)) {
+        errorMessages.push('port id is required');
+      }
+    });
+
+    return errorMessages;
+  }
+
+  _isValidPortId(id) {
+    return id !== null && id !== undefined && !isObject(id);
```

Comment:
We might not need this kinda check, because the Browser Protocol layer will do the validation for us.